README.md
# ParaDict

![ParaDict](/public/og-image.png)

ParaDict is a single web page application for searching multiple words meaning more easily, by search for the words in parallel on [Longdo Dictionary](https://dict.longdo.com/) so that all search entries are returned immediately. It has a clean and minimal interface for ease of use.

**This website is made for educational purposes only.** The source code of this website is licensed in MIT. The dictionaries used as a result from this website are licensed to their respective owners.

## Features
- Search for multiple words at once.
- Easily bookmark the search results for later access by saving the results URL.
- Returns an error when the given search entry is not found, and allow editing or removing the entry from the current result list.
- Runs at the Edge for faster response time!

## Technologies
- **Next.js** for web stack.
- **HeadlessUI** for complex UI components.
- **TailwindCSS** for styling.
- **Remark** for parsing autocomplete markdown.
- **LinkedDOM** for parsing HTML results from Longdo Dictionary API.

CHANGELOG.md
# ParaDict ChangeLog

## v2.2
- fix: improve Hope Dictionary parsing for multiple part of speech definitions

## v2.1 (20230712)
- refactor: change query param from `search` to `q`
- feat: add edit all search entries

## v2.0 (20230624)
- refactor: update all dependencies
- refactor: rewrite API for edge runtime
- feat: edit and delete invalid search entries after submit

## v1.3 (20220123)
- feat: Use cached results from recent words.

## v1.2 (20220108)
- feat: Add autocomplete function.
- fix: definition with `See also` field not parsing properly.

## v1.1 (20211225)
- feat: Add parsing from Hope Dictionary.
- feat: Design improvements.
- feat: Add queing system.

## v1.0 (20211224)
- Initial release

shared/search-query.ts
import { NextRouter } from 'next/router'
import type { ParsedUrlQuery } from 'querystring'
import { WordsRequest } from '../types/dict'

export const SEARCH_QUERY_KEY = 'q'
export const EDIT_QUERY_KEY = 'edit'

export const normalizeQueryArray = (query: string | string[]): string[] => {
  if (Array.isArray(query)) return query
  return [query]
}

export const shallowReplace = (router: NextRouter, targetQuery: ParsedUrlQuery | null) => {
  router.replace(
    {
      pathname: router.pathname,
      ...(targetQuery
        ? {
            query: targetQuery,
          }
        : {}),
    },
    undefined,
    {
      shallow: true,
    }
  )
}

export const isSearchView = (
  query: ParsedUrlQuery
): query is {
  [SEARCH_QUERY_KEY]: string[]
} => {
  if (query?.[EDIT_QUERY_KEY]) return false
  const searchQuery = query?.[SEARCH_QUERY_KEY]
  if (!searchQuery) return false
  return normalizeQueryArray(searchQuery).length > 0
}

export const shouldRefactorSearchQuery = (query: ParsedUrlQuery): ParsedUrlQuery | undefined => {
  const v1_searchQuery = query?.search
  if (!v1_searchQuery) return
  const filteredQuery: ParsedUrlQuery = {
    ...query,
    [SEARCH_QUERY_KEY]: normalizeQueryArray(v1_searchQuery),
  }
  if (filteredQuery.search) delete filteredQuery.search
  return filteredQuery
}

export const convertQueryToWordsRequest = (query: ParsedUrlQuery): WordsRequest[] => {
  const searchQuery = query?.[SEARCH_QUERY_KEY]
  if (!searchQuery) return []
  return normalizeQueryArray(searchQuery).map((w) => ({
    search: w,
  }))
}

export const convertWordsRequestToQuery = (wordsRequest: WordsRequest[]): ParsedUrlQuery => {
  return {
    [SEARCH_QUERY_KEY]: wordsRequest
      .filter((word) => word.search && typeof word.search === 'string')
      .map((word) => word.search),
  }
}

export const toggleEditQuery = (query: ParsedUrlQuery): ParsedUrlQuery => {
  const editQuery = query?.[EDIT_QUERY_KEY]
  if (editQuery) {
    const filteredQuery = { ...query }
    delete filteredQuery[EDIT_QUERY_KEY]
    return filteredQuery
  }
  return {
    ...query,
    [EDIT_QUERY_KEY]: 'true',
  }
}


pages/api/autocomplete.ts
import { NextRequest, NextResponse } from 'next/server'
import ky from 'ky'
import { SelectOption } from '../../types/select'

type RawResponse = {
  s: 'head'
  /** Formatted HTML word string */
  d: string
  /** Actual Word */
  w: string
  id: string
}

export const config = {
  runtime: 'edge',
}

const processLabel = (r: RawResponse) => {
  return r.d.replace(/(<FONT color=blue>)|(<\/FONT>)/g, '**').replace(/\*\*\*\*/g, '')
}
const handler = async (req: NextRequest) => {
  const params = req.nextUrl.searchParams
  const search = params.get('search')
  if (
    typeof search !== 'string' ||
    // Only allow a search with more than 3 characters
    search.length < 3 ||
    req.method !== 'GET'
  )
    return NextResponse.json({ error: 'Bad Request' }, { status: 400 })

  //https://search.longdo.com/BWTSearch/HeadSearch?json=1&ds=head&num=20&count=5&key=extra
  const data = await ky
    .get('BWTSearch/HeadSearch', {
      prefixUrl: 'https://search.longdo.com',
      searchParams: {
        json: '1',
        ds: 'head',
        num: '20',
        count: search.length,
        key: search,
      },
    })
    .text()
  const results: RawResponse[] = JSON.parse(
    data.slice(data.indexOf('['), data.lastIndexOf(']') + 1).replace(/\r?\n|\r/g, '')
  )

  const resultsMap = new Map<string, SelectOption>()
  results.forEach((r) => {
    const label = processLabel(r)
    resultsMap.set(label.toLowerCase(), {
      label: label.toLowerCase(),
      value: r.w.toLowerCase(),
    })
  })

  return NextResponse.json<SelectOption[]>(Array.from(resultsMap.values()), {
    status: 200,
    headers: {
      'cache-control': `public, max-age=${86400 / 2}, stale-while-revalidate=86400`,
    },
  })
}
export default handler


shared/axios-concurrency.ts
import { AxiosInstance } from 'axios'

/**
 * Axios Concurreny Library, but typescript.
 *
 * See https://github.com/bernawil/axios-concurrency
 */
export const ConcurrencyManager = (axios: AxiosInstance, MAX_CONCURRENT = 10) => {
  if (MAX_CONCURRENT < 1) throw 'Concurrency Manager Error: minimum concurrent requests is 1'
  let instance = {
    queue: [],
    running: [],
    shiftInitial: () => {
      setTimeout(() => {
        if (instance.running.length < MAX_CONCURRENT) {
          instance.shift()
        }
      }, 0)
    },
    push: (reqHandler) => {
      instance.queue.push(reqHandler)
      instance.shiftInitial()
    },
    shift: () => {
      if (instance.queue.length) {
        const queued = instance.queue.shift()
        queued.resolver(queued.request)
        instance.running.push(queued)
      }
    },
    // Use as interceptor. Queue outgoing requests
    requestHandler: (req) => {
      return new Promise((resolve) => {
        instance.push({ request: req, resolver: resolve })
      })
    },
    // Use as interceptor. Execute queued request upon receiving a response
    responseHandler: (res) => {
      instance.running.shift()
      instance.shift()
      return res
    },
    responseErrorHandler: (res) => {
      return Promise.reject(instance.responseHandler(res))
    },
    interceptors: {
      request: null,
      response: null,
    },
    detach: () => {
      axios.interceptors.request.eject(instance.interceptors.request)
      axios.interceptors.response.eject(instance.interceptors.response)
    },
  }
  // queue concurrent requests
  instance.interceptors.request = axios.interceptors.request.use(instance.requestHandler)
  instance.interceptors.response = axios.interceptors.response.use(
    instance.responseHandler,
    instance.responseErrorHandler
  )
  return instance
}


pages/api/dict.ts
import { NextRequest, NextResponse } from 'next/server'
import { parseHTML } from 'linkedom/worker'
import ky from 'ky'
import { Dictionary, SearchAPIResult } from '../../types/dict'

export const config = {
  runtime: 'edge',
}

type HopeDictEntry = {
  partOfSpeech: string
  meaning: string
}

/**
 * Hope Dictionary returns a very long definition string.
 * Seperate and format it ourselves.
 *
 * However, since the raw results from the API itselves is not strictly formatted, it may parsed incorrectly and return some unexpected results.
 */
const processHopeDict = (definition: string): string[] => {
  const partOfSpeech = /([a-z])+[.,]?/
  const thaiPronounciation = /\(([ก-๏].)+\)/
  const results = definition.split(' ').reduce((data, define) => {
    // The current index of each part of speech definition. Defaults to the previous item.
    let currentIndex = data.length - 1
    if (partOfSpeech.test(define)) {
      // Part of speech indicates that this is a new definition.
      // If the previous item already has a meaning, we create a new item.
      if (data[currentIndex]?.meaning || currentIndex === -1) {
        data.push({
          partOfSpeech: '',
          meaning: '',
        })
        currentIndex = currentIndex + 1
      }
      // Concat the part of speech to the current item.
      data[currentIndex].partOfSpeech = data[currentIndex].partOfSpeech + define
    } else if (!thaiPronounciation.test(define) && data[currentIndex]?.partOfSpeech) {
      // Concat the meaning to the current item, excluding Thai pronounciation.
      data[currentIndex].meaning = data[currentIndex].meaning + ' ' + define
    }
    return data
  }, [] as HopeDictEntry[])
  return results.map((d) => `(${d.partOfSpeech}) ${d.meaning}`)
}

const isAllowedLanguage = (d: string[]) => {
  const english = /([a-zA-Z])+/g
  const thai = /([ก-๏])+/g
  // Adding EN-TH dictionary will result in too large payload.
  // Uncomment the condition if you know what you are doing!
  return (
    english.test(d[0]) && thai.test(d[1]) //|| (english.test(d.meaning) && thai.test(d.word))
  )
}

/**
 * Process each defintion by removing `See also:` and `Syn.` records.
 * @param define The definition string
 */
const processDefinition = (define: string) => {
  for (let _kw of ['See also:', 'Syn.']) {
    // Add seperator
    const kw = `, ${_kw}`
    if (define.indexOf(kw) !== -1) return define.slice(0, define.indexOf(kw)).trim()
  }
  return define.trim()
}
const processTable = (table: HTMLTableElement, baseWord: string) => {
  const rows = table.querySelectorAll('tr')
  return Array.from(rows)
    .map((row) => {
      const cells = row.querySelectorAll('td')
      return Array.from(cells).map((cell, i) =>
        i === 1 ? processDefinition(cell.textContent) : cell.textContent
      )
    })
    .filter((d) => baseWord.toLowerCase() === d[0].toLowerCase() && isAllowedLanguage(d))
}

const isHTMLTable = (elem: Element): elem is HTMLTableElement => {
  return elem.tagName === 'TABLE'
}

const handler = async (req: NextRequest) => {
  const searchParams = req.nextUrl.searchParams
  const search = searchParams.get('search')
  if (typeof search !== 'string' || req.method !== 'GET')
    return NextResponse.json({ error: 'Invalid request' }, { status: 400 })
  const data = await ky
    .get('mobile.php', {
      prefixUrl: 'https://dict.longdo.com',
      searchParams: {
        search,
      },
    })
    .text()
  const { document } = parseHTML(data.replace(/\t/g, '').replace(/ {2}/g, ' '))
  const children = document.querySelectorAll('body > *')
  const dictResults: Dictionary[] = []
  for (var i = 0; i < children.length; i++) {
    const elem = children[i]
    if (isHTMLTable(elem) && elem.classList.contains('result-table')) {
      const dict = children[i - 1].textContent
      const results = processTable(elem, search)
      if (results.length === 0) continue
      dictResults.push({
        dict,
        results: dict.includes('Hope') ? processHopeDict(results[0][1]) : results.map((d) => d[1]),
      })
    }
  }

  return NextResponse.json<SearchAPIResult>(
    {
      word: search,
      data: dictResults,
    },
    {
      status: dictResults.length === 0 ? 404 : 200,
      headers: {
        'cache-control': `public, max-age=${86400 / 2}, stale-while-revalidate=86400`,
      },
    }
  )
}

export default handler
